pdfiumandroid/src/main/cpp/pdfiumandroid.cpp

```diff
@@ -58,12 +58,41 @@ struct rgb {
     uint8_t blue;
 };
 
+JavaVM* javaVm;
+
+bool jniAttachCurrentThread(JNIEnv **env, bool *attachedOut) {
+    JavaVMAttachArgs jvmArgs;
+    jvmArgs.version = JNI_VERSION_1_6;
+
+    bool attached = false;
+    if (javaVm->GetEnv((void **) env, JNI_VERSION_1_6) == JNI_EDETACHED) {
+        if (javaVm->AttachCurrentThread(env, &jvmArgs) != JNI_OK) {
+            LOGE("Cannot attach current thread");
+            return false;
+        }
+        attached = true;
+    } else {
+        attached = false;
+    }
+    *attachedOut = attached;
+    return true;
+}
+
+bool jniDetachCurrentThread(bool attached) {
+    if (attached && javaVm->DetachCurrentThread() != JNI_OK) {
+        LOGE("Cannot detach current thread");
+        return false;
+    }
+    return true;
+}
+
 class DocumentFile {
 
 public:
     FPDF_DOCUMENT pdfDocument = nullptr;
 
-    public:
+public:
+    jobject nativeSourceBridgeGlobalRef = nullptr;
```

Comment:
Here's another challenge. We want to store some pointer to the object that we're going to use for read callback. That pointer is a reference to Java object, which gets passed as one of the arguments to the methods. However, just like JNIEnv, the reference to such object will be invalid outside of the method call. Therefore, it has to be wrapped with `NewGlobalReference()`. However, this reference also needs to be released once the document is closed. I did it in the similar way to the cDataCopy: in fact, this is a pretty common pattern for JNI wrappers

Reply:
That's fine (that we'll need to clean it up on document close).  We are already do that for other things.  We might want to have some sort of registry system for things that need to be called when the document is closed.